Adds workflow approval list and details

mabashian opened the pull request:
##### SUMMARY
Meta issue https://github.com/ansible/awx/issues/8273
Addresses #8373 
Sort of addresses https://github.com/ansible/awx/issues/7112 (the rest of it captured in https://github.com/ansible/awx/issues/8374)
Does not include https://github.com/ansible/awx/issues/8372

This PR adds a new list `/workflow_approvals` along with a details view for items in that list.

The entry point for this list is here:

<img width="1680" alt="Screen Shot 2020-10-12 at 2 36 43 PM" src="https://user-images.githubusercontent.com/9889020/95779699-7c9a9580-0c98-11eb-95e0-ebb06612374e.png">

Since this list does not exist in the old UI I figured this was as good a spot as any.  We can move it if we need to.

The list itself is pretty straightforward but here's an explanation of the columns:

<img width="1458" alt="Screen Shot 2020-10-12 at 2 45 12 PM" src="https://user-images.githubusercontent.com/9889020/95780268-9688a800-0c99-11eb-9793-cbd5b0a716ee.png">

<img width="1449" alt="Screen Shot 2020-10-12 at 2 47 18 PM" src="https://user-images.githubusercontent.com/9889020/95780412-dea7ca80-0c99-11eb-9943-5c05417b236c.png">

<img width="1446" alt="Screen Shot 2020-10-12 at 2 50 26 PM" src="https://user-images.githubusercontent.com/9889020/95780664-4cec8d00-0c9a-11eb-9bb5-1ed061c6690a.png">

Some examples of tooltip on status:

<img width="388" alt="Screen Shot 2020-10-12 at 2 51 00 PM" src="https://user-images.githubusercontent.com/9889020/95780727-67bf0180-0c9a-11eb-93b1-10de29827f57.png">
<img width="400" alt="Screen Shot 2020-10-12 at 2 51 06 PM" src="https://user-images.githubusercontent.com/9889020/95780729-68579800-0c9a-11eb-8f31-2f56bae8872d.png">

<img width="1452" alt="Screen Shot 2020-10-12 at 2 54 38 PM" src="https://user-images.githubusercontent.com/9889020/95780960-e1ef8600-0c9a-11eb-88d0-baa1b980e96c.png">

Here's a look at the details view for a pending approval:

<img width="1435" alt="Screen Shot 2020-10-12 at 2 55 15 PM" src="https://user-images.githubusercontent.com/9889020/95781001-f59aec80-0c9a-11eb-8646-29fb1578b0e0.png">

a lot of the same information from the list row is shown here.

Here's a look at the details view for an approved approval:

<img width="1412" alt="Screen Shot 2020-10-12 at 2 56 11 PM" src="https://user-images.githubusercontent.com/9889020/95781085-19f6c900-0c9b-11eb-8cb1-20e154fee9f6.png">

The action buttons/rbac should be the same on the details view as they are on the list.

**Websockets**

OK so here's how websockets work on this list.  This list is subscribed to the jobs -> status_changed channel just like the jobs list.  We're only interested in jobs with the type: `workflow_approval` though.  The response to events that match our criteria is pretty straightforward: we refresh the list contents.  This is throttled but that throttle is currently set to just 1 second.  Unlike the jobs list, I'm not anticipating a tremendous amount of activity on this list in a short amount of time but if we need to increase that throttle threshold then we can.  Here's when we refresh the list:

1) When the job in question _is not_ in our list _and_ it's status is `pending`.
2) When the job in question _is_ in our list _and_ it's status is not in `['new', 'pending', 'waiting', 'running']` (i.e. the job finished).

^^ Scenario 2 creates an interesting challenge when the user that's viewing the list is also the user that is acting on the row.  When the user clicks approve/deny and the API responds with a 204 there's a bit of a delay before the socket message comes through and tells us that the job has moved to a completed status.  During that time, I hide the action buttons via a boolean on the list row state.

**Websockets are where I spent a majority of my time on this feature and I'm still not sure that it's quite right.  I'm definitely open to alternative approaches to keeping this list up to date using the data that we get from the job status channel.**

Unfortunately, we have to hit the API to retrieve the row data as the socket message doesn't have much more than the status and the job id.  There could be a solution where we fetch specific rows instead of refetching all of them.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

mabashian committed:
Adds /#/workflow_approvals list and details and allows users to approve or deny workflow approvals from these interfaces

wenottingham commented:
Looks generally good. I do think we need to move the list - workflow approvals are the same conceptual things as job runs, and so should be in that area.

unlikelyzero commented:
@tiagodread note: This is net new and not absolutely critical for this week.

mabashian committed:
Move approval action buttons from rows to to level list actions.  UX updates to the display of the status.

mabashian commented:
OK, made the following updates based on feedback from Taufique and Bill.

1) Moved Workflow Approvals out of Administration and in to Views
2) Moved approve/deny buttons from the individual list row items up to the list action level (alongside delete).  This means multiple rows can be approved/denied at the same time.
3) Updated the UX on the status labels to match the mockups: https://tower-mockups.testing.ansible.com/patternfly/workflow-approvals/wf-approvals/

jakemcdermott reviewed (commented):
- List item, approval button, new view, and tests, etc. LGTM
- let's double-check w/ @keithjgrant on the websocket hook usage

keithjgrant commented on awx/ui_next/src/screens/WorkflowApproval/shared/WorkflowApprovalStatus.jsx:
```diff
@@ -0,0 +1,69 @@
+import { WorkflowApproval } from '../../../types';
+import { formatDateString } from '../../../util/dates';
+
+function WorkflowApprovalStatus({ workflowApproval, i18n }) {
```
We have a generic status label component at `components/StatusLabel`. Do you think it could meet your needs here instead of a new component?

mabashian replied:
👍 will check that out

mabashian replied:
Yea there's definitely some overlap here.  If we want to keep the same UX then I'd have to refactor this component quite a bit with several conditionals.  I'm a little reluctant to do that given how clean this component (StatusLabel) is.

AlexSCorey commented on awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalDetail/WorkflowApprovalDetail.jsx:
```diff
@@ -0,0 +1,246 @@
+          <ErrorDetail error={deleteError} />
+        </AlertModal>
+      )}
+      {approveError && (
```
You could combine these 2 AlertModals, and would you also only need to have 1 useDismissableError above, which would be something like
```const {error, dismissError} = useDismissableError(approveError || denyError)```
and then the alertModals would be
```<AlertModal
          isOpen={error}
          variant="error"
          title={i18n._(t`Error!`)}
          onClose={dismissError}
        >
         approveError ?  {i18n._(t`Failed to approve workflow approval.`)} 
            : {i18n._(t`Failed to deny workflow approval.`)}
          <ErrorDetail error={approveError} />
        </AlertModal>
```

AlexSCorey commented on awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalList.jsx:
```diff
@@ -0,0 +1,277 @@
+    fetchWorkflowApprovals();
+  }, [fetchWorkflowApprovals]);
+
+  // TODO: update QS_CONFIG to be safe for deps array
```
Is this a note for your current work that you forgot to remove, or is this future work?

keithjgrant replied:
that's an old TODO of mine, copied over from somewhere else I think

AlexSCorey commented on awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalList.test.jsx:
```diff
@@ -0,0 +1,317 @@
+
+jest.mock('../../../api');
+
+const mockWorkflowApprovals = [
```
Might consider putting this inside of a data.json file to reduce the # of lines of code in this file

mabashian replied:
👍

AlexSCorey commented on awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/useWsWorkflowApprovals.js:
```diff
@@ -0,0 +1,61 @@
+import { useState, useEffect } from 'react';
```
was does the Ws in `useWsWorkflowApprovals` stand for?

mabashian replied:
That's the naming pattern we have for all of our websocket hooks.  They all start with `useWs` and then we concat the resource on the end.

tiagodread commented:
Hey @mabashian !
- I tried to delete a workflow approval before approving or deny and I got this result (on the list or details):

![approval](https://user-images.githubusercontent.com/12464858/97048385-b600ba00-1550-11eb-9d50-0abbaab367d5.gif)

- The simple search using name doesn’t work:
![search](https://user-images.githubusercontent.com/12464858/97049129-89996d80-1551-11eb-9068-0c9d6b9be8b6.gif)

- If i delete the job the workflow approval associated with that job show `undefined` for the job name and if I click on it I got 404:

![deleted](https://user-images.githubusercontent.com/12464858/97049368-fca2e400-1551-11eb-9e27-8410bf04e7fb.gif)

unlikelyzero commented on awx/ui_next/src/screens/WorkflowApproval/WorkflowApprovalList/WorkflowApprovalListApproveButton.jsx:
```diff
@@ -0,0 +1,76 @@
+          <div>
+            <Button
+              isDisabled={isDisabled}
+              aria-label={i18n._(t`Approve`)}
```
do aria-labels need to be translated?

unlikelyzero commented:
We have tests pending the merge of this PR. Please let us know when it's ready for final merge

tiagodread commented:
@mabashian can you add some uniques ids for the data on datails tab specially for the status

mabashian committed:
Combine approve/deny modals into one modal

mabashian committed:
Remove old TODO

mabashian committed:
Move all mock workflow approvals out to data.workflowApprovals.json

mabashian committed:
Prevent users from attempting to delete pending approvals

mabashian committed:
Fix basic name filtering on workflow approvals list

mabashian committed:
Add description to the list of default search keys on workflow approvals list

mabashian committed:
Show 'Deleted' when the source workflow job is deleted in workflow approvals list/details

mabashian committed:
Adds data-cy attributes to the fields in the workflow approval details

mabashian commented:
OK @tiagodread @unlikelyzero I think this is ready for your eyes again.  I've added data-cy attributes to all the details: https://github.com/ansible/awx/pull/8375/files#diff-eeea31a09935ea3101ae734da15ca14d933119b4bb0819194fb9ff24a074268cR85-R183

unlikelyzero commented:
@mabashian hey, we're not longer using data-cy attributes but trying to use the dataouia attributes instead. Is there enough to glean from this PR?

https://github.com/ansible/awx/pull/8472#discussion_r517504241

tiagodread commented:
Hey @mabashian looks like the error `- I tried to delete a workflow approval before approving or deny and I got this result (on the list or details)` was fixed for the list, but, i can try to delete a pending workflow approval in the details page:

![wfaproval](https://user-images.githubusercontent.com/12464858/99297478-12d95280-2827-11eb-8726-1470d6b7cb53.gif)

mabashian committed:
Hide the delete button on workflow approval details when job is pending

mabashian committed:
Update snapshot after adding cannotDelete prop to ToolbarDeleteButton

mabashian committed:
Wrap status label in a div so that it doesn't vertically stretch with it's parent.

tiagodread commented:
@mabashian awesome job!!!

Tests covered by a mix of manual and automated test cases in here https://github.com/ansible/tower-qa/pull/5656

mabashian commented:
@keithjgrant @jakemcdermott @AlexSCorey any other thoughts on this one before merge?
